homeassistant/components/compensation/__init__.py

```diff
@@ -0,0 +1,121 @@
+"""The Compensation integration."""
+import logging
+import warnings
+
+import numpy as np
+import voluptuous as vol
+
+from homeassistant.components.sensor import DOMAIN as SENSOR_DOMAIN
+from homeassistant.const import (
+    CONF_ATTRIBUTE,
+    CONF_ENTITY_ID,
+    CONF_NAME,
+    CONF_UNIT_OF_MEASUREMENT,
+)
+from homeassistant.helpers import config_validation as cv
+from homeassistant.helpers.discovery import async_load_platform
+
+from .const import (
+    CONF_COMPENSATION,
+    CONF_DATAPOINTS,
+    CONF_DEGREE,
+    CONF_POLYNOMIAL,
+    CONF_PRECISION,
+    DATA_COMPENSATION,
+    DEFAULT_DEGREE,
+    DEFAULT_NAME,
+    DEFAULT_PRECISION,
+    DOMAIN,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def datapoints_greater_than_degree(value: dict) -> dict:
+    """Validate data point list is greater than polynomial degrees."""
+    if len(value[CONF_DATAPOINTS]) <= value[CONF_DEGREE]:
+        raise vol.Invalid(
+            f"{CONF_DATAPOINTS} must have at least {value[CONF_DEGREE]+1} {CONF_DATAPOINTS}"
+        )
+
+    return value
+
+
+COMPENSATION_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_ENTITY_ID): cv.entity_id,
+        vol.Required(CONF_DATAPOINTS): [
+            vol.ExactSequence([vol.Coerce(float), vol.Coerce(float)])
+        ],
+        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
```

Comment:
Should we allow setting a unique id instead? Then the user can change name and entity_id in the GUI.

Reply:
Yeah I can do that.  I think I originally made this before I knew about unique_id.

Reply:
I didn't mean that we should add the name option back, just that we should set a better default name than unnamed device.

Reply:
I'll do compensation source and attribute, that's what others do. Sound good?

Reply:
Yes, sounds good!

Reply:
So, should I grab the friendly name from the source or use the entity_id?  Right now the code is using the entity_id, but before I commit I wanted your thoughts.

Reply:
or is there a helper similar to async_generate_entity_id that's for friendly names?

Reply:
Just return the name in the `name` property of the entity.